internal/testprotos/make_protos.sh

```diff
@@ -4,15 +4,34 @@ set -e
 
 cd $(dirname $0)
 
+PROTOC_VERSION="3.5.1"
+PROTOC_OS="$(uname -s)"
+PROTOC_ARCH="$(uname -m)"
+case "${PROTOC_OS}" in
+  Darwin) PROTOC_OS="osx" ;;
+  Linux) PROTOC_OS="linux" ;;
+  *)
+    echo "Invalid value for uname -s: ${PROTOC_OS}" >&2
+    exit 1
+esac
+
+rm -rf ./protoc
```

Comment:
Instead of always removing and re-downloading, it would be really nice to just use what's there if it's right.

So I think a few more steps make sense:
1. See if `protoc` is already on the path: `hash protoc`
2. If it is, run `protoc --version` and see if the output matches the desired version. If it does *done*: `PROTOC=protoc`
3. If it's not on the path or it spits out the wrong version, look for `./protoc/bin/protoc`. If it exists, try `protoc --version` for it. If it's the right version, *done*. `PROTOC=./protoc/bin/protoc`
4. Finally, run the logic here to download the right version. After downloading, sanity check it with `protoc --version` and fail if it somehow has the wrong output.

Reply:
I'm happy to do this, just wanted to put my two cents in first and confirm - the download/unzip takes on the order of seconds, so instead of (IMO) overcomplicating this script, I figured the easiest way to make sure we always have a consistent state is to just run the steps again. This also guarantees determinism relative to the WKT includes - a developer could theoretically edit one of the files in `protoc/include/google/protobuf` and then run this, and the above steps won't pick up that the file changed. By redoing the download before generating, we guarantee this isn't an issue. Does that make sense?

Reply:
I have a "trust on first use" kind of mentality -- I don't like the idea of downloading code from the internet *every single time* I want to do a task. But I think we can come to a happy compromise:

1. Ignore what's on the path. Only trust what's in `./protoc`.
2. Instead of asking for its version, compute a SHA for the whole directory.
3. Hardcode the expected SHA in this script. If the SHA of the directory does not match, then download it.

To compute the SHA for a whole directory, I think you could do this:
```bash
find protoc -type f | sort | xargs shasum | shasum
```
But this relies on the output of `shasum` being consistent on mac and linux. (I think it's called `sha1sum` on linux.) You might include two known SHAs and switch on OS like you already have at the top.

WDYT?

Reply:
Heh, alternatively I can just do what you suggest - it's your repo after all :-) No need to overcomplicate it, and if you want to trust it, totally fine. Will update soon.

Reply:
I would highly not recommend converting this to bazel or pants, way too heavy for this repo IMO :-) I'll update in a bit.

Reply:
LOL. Fair enough. Though it's just a standard Go repo, so the Go tool works with it regardless. I guess I just meant wiring up the code gen steps and all dependencies with a smarter build tool. But, in any event, I agree -- not worth the hassle and custom scripts are fine. Especially since there are few people submitting PRs against it, and these things are only ever used by people editing the code, not people using it.

Reply:
OK so here's what I'm going to do: I want to at least use always use a downloaded copy, so that I can rely on having the `include/google/protobuf` directory and there's no weirdness, so I'll check to see if `protoc/bin/protoc` exists and outputs the correct version, otherwise I'll re-download.

Reply:
Updated